product_profile/models/product.py

```diff
@@ -0,0 +1,286 @@
+    model = type(self)._name
+    message = (_("Issue\n------\n"
+                 "%s\n'%s' value can't be applied to '%s' field."
+                 "\nThere is no matching value between 'Product Profiles' "
+                 "\nand '%s' models for this field.\n\n"
+                 "Resolution\n----------\n"
+                 "Check your settings on Profile model:\n%s"
+               % (error, value, field, model, PROFILE_MENU)))
+    return message
+
+
+def get_profile_fields_to_exclude():
+    # These fields must not be synchronized between product.profile
+    # and product.template/product
+    return models.MAGIC_COLUMNS + [
+        'name', 'explanation', 'sequence',
+        'display_name', '__last_update']
+
+
+class ProductProfile(models.Model):
+    _name = 'product.profile'
+    _order = 'sequence'
+
+    name = fields.Char(
+        required=True,
+        help="Profile name displayed on product template\n"
+             "(not synchronized with product.template fields)")
+    sequence = fields.Integer(
+        help="Defines the order of the entries of profile_id field\n"
+             "(not synchronized with product.template fields)")
+    explanation = fields.Text(
+        required=True,
+        oldname='description',
+        help="An explanation on the selected profile\n"
+             "(not synchronized with product.template fields)")
+    type = fields.Selection(
+        selection='_get_types',
+        required=True,
+        help="See 'type' field in product.template")
+
+    def _get_types(self):
+        """ inherit in your custom module.
+            could be this one if stock module is installed
+
+        return [('product', 'Stockable Product'),
+                ('consu', 'Consumable'),
+                ('service', 'Service')]
+        """
+        return [('consu', 'Consumable'), ('service', 'Service')]
+
+    @api.multi
+    def write(self, vals):
+        """ Profile update can impact products: we take care
+            to propagate ad hoc changes """
+        new_vals = vals.copy()
+        excludable_fields = get_profile_fields_to_exclude()
+        for key in vals:
+            if (key.startswith(PROF_DEFAULT_STR) or
+                    key in excludable_fields or
+                    self.check_useless_key_in_vals(new_vals, key)):
+                new_vals.pop(key)
+        # super call must be after check_useless_key_in_vals() call
+        # because we compare value before and after write
+        res = super(ProductProfile, self).write(new_vals)
+        if new_vals:
+            for rec in self:
+                products = self.env['product.product'].search(
+                    [('profile_id', '=', rec.id)])
+                if products:
+                    _logger.info(
+                        " >>> %s Products updating after updated '%s' "
+                        "product profile" % (len(products), rec.name))
+                    products.write({'profile_id': rec.id})
```

Comment:
it is to force product update after  profile change
see next commit for update code